### ISSUE TITLE
Allow `oxc_parser` to consume / update an existing sourcemap that `oxc_codegen` can then output

### ISSUE THREAD

**Issue by 0xdevalias** [oxc-project/oxc#23064]
It would be useful if `oxc_parser` was able to consume an existing sourcemap alongside the JS being parsed, and 'update' it as we transform the AST, which `oxc_codegen` can then output; in a way that would allow the 'final end result' to be mapped all the way back to the original source.

From a quick exploration of the docs/etc, I don't think this is currently possible.

I also saw this comment from 2024 implying it wasn't possible then at the very least:

- https://github.com/oxc-project/oxc/issues/7461#issuecomment-2496828064
  - > Or... implement support for input sourcemaps in Oxc parser (which would be good to have anyway).

I know [`oxc_codegen`](https://crates.io/crates/oxc_codegen) can currently emit sourcemaps on it's own; and I can see there is an [`oxc_sourcemap`](https://crates.io/crates/oxc_sourcemap), and that has a `ConcatSourceMapBuilder`, but from a quick read I don't think that actually does what I am asking for here:

- https://docs.rs/oxc_sourcemap/latest/oxc_sourcemap/struct.ConcatSourceMapBuilder.html

So I guess it might also be useful to have a 'lower level support' in `oxc_sourcemap` as well.

---

As a piece of 'prior art', if I am reading it correctly, Parcel's `source-map` library seems to allow combining sourcemaps as proposed here:

- https://github.com/parcel-bundler/source-map#creating-from-existing-sourcemap
- https://parceljs.org/plugin-system/source-maps/#transformations%2Fmanipulations

## See Also

- https://github.com/oxc-project/oxc/issues/23065

**Comment by Boshen** [comment]
Do you have a specific real-world use case that would help me understand?

**Comment by 0xdevalias** [comment]
> Do you have a specific real-world use case that would help me understand?

@Boshen At a high level:

- If I have original source `A`
- That gets transformed (potentially by some other tool) into `B` + `BSourceMap`
- That I then want to parse with `oxc_parser`, do my own AST transforms, and then `oxc_codegen` into `C` + `CSourceMap`

Such that debugging `C` + `CSourceMap` can correctly point me back to the correct place in `A`

Essentially a typical bundler/minifier sort of use case, e.g:

- https://parceljs.org/plugin-system/source-maps/#transformations%2Fmanipulations

Is that concrete/'real-world' enough? Or did you want me to try and describe it less abstractly?

---

Currently a more specific form of that would be in code formatted with `oxc_formatter` (see https://github.com/oxc-project/oxc/issues/23065) as that is the only part the project I'm contributing to currently uses; but I am also exploring a deeper integration with the `oxc` ecosystem for it as well (see https://github.com/pionxzh/wakaru/issues/152) where this would become even more useful.

That project is essentially like a bundler/minifier, but in reverse. Based on the current architecture it:

- Parses a `.js` file
- Detects bundle semantics and module boundaries, extracts + normalises the modules, and emits standalone module strings
- Then in parallel:
  - those module strings are re-parsed
  - a number of AST transform rules are run on them
  - they are codegen'd to strings
  - Those strings are then re-parsed+formatted
  - And written to output files

As one example in this, being able to consume existing SourceMaps would help 'cross the boundaries' where it serialises back to strings (particularly as different steps use different parsers/etc currently); though there are likely other ways to work around that part more manually as well.

Then, while it would be out of scope for the specifics in this issue, the eventual goal would then to generate a 'reverse source map' from that unminified/formatted `.js`; such that we could debug the original source `A` with that derived SourceMap:

- https://github.com/pionxzh/wakaru/issues/66

**Comment by 0xdevalias** [comment]
As an example of 'prior art', I asked GitHub Copilot (GPT-5.4) to explore how it's supported in `swc`'s ecosystem:

<details><summary>Summarising <code>swc</code>'s SourceMap support with GitHub Copilot (GPT-5.4)</summary>

## Prompt

```
Does swc (or particularly its lower level rust libraries) support input sourcemaps, transforms, and updated output sourcemaps?

Ideally I want to know how the parser/codegen/etc support that
```

With access to inspect the repo:

- https://github.com/swc-project/swc

## Response

> Yes — **swc does support the full “input sourcemap -> transform -> updated output sourcemap” flow**, and the lower-level Rust crates expose the relevant pieces too.
> 
> Short version:
> 
> - **Parsing** uses `swc_common::SourceMap` / `SourceFile` / spans to track original source locations.
> - **Transforms** mostly preserve/update those spans as AST nodes are rewritten.
> - **Codegen** (`swc_ecma_codegen::text_writer::JsWriter`) records generated-position ↔ span mappings.
> - **Final sourcemap generation** happens via `swc_common::SourceMap::build_source_map(...)`.
> - If you provide an **input/original sourcemap**, swc can **compose** the newly generated map with it via `orig.adjust_mappings(&map)`, producing an updated output map that points back through the previous transformation layer.
>
> ## How it works in swc
> 
> ### 1. Parsing: spans come from `swc_common::SourceMap`
> 
> The parser itself does **not** consume an input sourcemap. It parses source text into AST nodes with `Span`s backed by `swc_common::SourceMap` / `SourceFile`. That is the foundation for later sourcemap generation. See `parse_js(...)` and the common parser path here:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_compiler_base/src/lib.rs#L71-L125
> 
> And the lower-level source map builder API is defined in `swc_common`:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_common/src/source_map.rs#L1173-L1183
> 
> So: **parser = spans from source text, not direct sourcemap consumption**.
> 
> ---
> 
> ### 2. Input sourcemaps: supported at the compiler/high-level layer
> 
> The `swc` crate’s compiler explicitly supports loading an original/input sourcemap:
> 
> - from `//# sourceMappingURL=...`
> - from inline `data:` sourcemaps
> - from a user-provided sourcemap JSON string
> 
> That logic is in `Compiler::get_orig_src_map(...)`:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc/src/lib.rs#L429-L603
> 
> Then the high-level transform path threads that original map into printing:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc/src/lib.rs#L969-L985
> 
> And `PrintArgs` carries `orig: Option<swc_sourcemap::SourceMap>` all the way down:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_compiler_base/src/lib.rs#L127-L143
> 
> So **input sourcemaps are definitely supported**, but mainly via the higher-level compiler / minifier pipeline rather than the parser itself.
> 
> ---
> 
> ### 3. Codegen: records generated-to-original mappings
> 
> `swc_ecma_codegen`’s `JsWriter` accepts a mutable mapping buffer, and codegen populates it while emitting text.
> 
> Example from the repo:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_ecma_codegen/examples/sourcemap.rs#L32-L47
> 
> And the general print path does the same internally:
> 
> https://github.com/swc-project/swc/blob/main/crates/swc_compiler_base/src/lib.rs#L205-L223
> 
> So **codegen itself doesn’t directly “write a sourcemap file”**; instead it emits text while collecting mapping tuples, and `swc_common::SourceMap::build_source_map(...)` turns those into a real sourcemap.
> 
> ---
> 
> ### 4. Composition with an existing input map
> 
> This is the important bit for your question.
> 
> Inside `swc_common::build_source_map(...)`, swc first builds a map from generated output back to the spans/source files in the current AST. Then, if `orig` is present, it **adjusts/composes** the generated map against the input map:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_common/src/source_map.rs#L1421-L1428
> 
> That means the output sourcemap is not just “new generated → intermediate input”; it can be rewritten to point back through the original incoming sourcemap.
> 
> This is the core support for **input sourcemaps + transforms + updated output sourcemaps**.
> 
> ---
> 
> ### 5. The sourcemap crate itself is full-featured
> 
> `crates/swc_sourcemap` handles parsing and writing sourcemaps.
> 
> Decoding:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_sourcemap/src/decoder.rs#L151-L283
> 
> Encoding:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_sourcemap/src/encoder.rs#L149-L185
> 
> And `swc_config::SourceMapContent` can convert either raw JSON or parsed sourcemap fields into a `swc_sourcemap::SourceMap`:
> 
> - https://github.com/swc-project/swc/blob/main/crates/swc_config/src/source_map.rs#L29-L33
> 
> So at the lower-level crate layer, yes: **decode, manipulate, rewrite, and encode** are all supported.
> 
> ---
> 
> ## What “support” means at each layer
> 
> ### Parser
> 
> - **Supports location tracking via spans**
> - **Does not itself apply input sourcemaps**
> - Works from source text / `SourceFile`
> 
> ### Transform layer
> 
> - Operates on AST spans
> - If transforms preserve meaningful spans, sourcemaps stay good
> - Newly synthesized nodes may have synthetic/dummy spans unless the transform author sets better ones
> 
> ### Codegen
> 
> - Emits generated code
> - Collects generated-position mappings from spans
> - Can emit scopes metadata in some cases
> 
> ### Final sourcemap generation
> 
> - Builds sourcemap from emitted mappings
> - Can optionally compose with an original input sourcemap
> - Can inline sources content, emit names, columns, ignoreList, etc.

</details>

From that, it seems that 'input SourceMaps' aren't supported at the parser/codegen level of things, but more that when generating a SourceMap from the `mappings` that `swc_ecma_codegen`'s `JsWriter` creates, that can be composed with an existing SourceMap to adjust the mappings:

- https://github.com/swc-project/swc/blob/4e1f8326295932d77faa3d617a5b7cb8ba993a38/crates/swc_common/src/source_map.rs#L1285-L1290
- https://github.com/swc-project/swc/blob/4e1f8326295932d77faa3d617a5b7cb8ba993a38/crates/swc_common/src/source_map.rs#L1421-L1426

The underlying `adjust_mappings` function seems to exist in the `swc_sourcemap` crate:

- https://github.com/swc-project/swc/blob/4e1f8326295932d77faa3d617a5b7cb8ba993a38/crates/swc_sourcemap/src/lazy/mod.rs#L454-L462
- https://github.com/swc-project/swc/blob/4e1f8326295932d77faa3d617a5b7cb8ba993a38/crates/swc_sourcemap/src/types.rs#L948-L983

And then higher level API's sort of wrap/compose those 'primitives' into nicer usage:

- https://github.com/swc-project/swc/blob/main/crates/swc/src/lib.rs#L429-L604
- https://github.com/swc-project/swc/blob/main/crates/swc/src/lib.rs#L970-L980
- https://github.com/swc-project/swc/blob/main/crates/swc_compiler_base/src/lib.rs#L127-L143

---

As another piece of 'prior art', if I am reading it correctly, Parcel's `source-map` library seems to allow combining sourcemaps as well:

- https://github.com/parcel-bundler/source-map
  - > Parcel's source-map library
  - > A source map library purpose-build for the Parcel bundler with a focus on fast combining and manipulating of source-maps.
- https://github.com/parcel-bundler/source-map#creating-from-existing-sourcemap
- https://parceljs.org/plugin-system/source-maps/#transformations%2Fmanipulations
  - ```js
    // The `extends` function uses the provided map to remap the original 
    // source positions of the map it is called on. In this case, the 
    // original source positions of `map` get remapped to the positions 
    // in `originalMap`.
    map.extends(originalMap);
    ```
  - > `extends(buffer: Buffer | SourceMap): SourceMap,`
    > 
    > Remaps original positions from this map to the ones in the provided map\
    > This works by finding the closest generated mapping in the provided map to original mappings of this map and remapping those to be the original mapping of the provided map.
    > 
    > **Params:**
    > 
    > - **`buffer`**: exported [SourceMap](https://parceljs.org/plugin-system/source-maps/#SourceMap) as a buffer
  - https://github.com/parcel-bundler/source-map/blob/b290dae1bea9bf249a0691e06014e09f974f70dd/parcel_sourcemap/src/lib.rs#L437-L515
    - This is the rust implementation of `extends`, which is summarised by GitHub Copilot (GPT-5.4) as:
      - > So, precisely, `extends` does this:
        > 
        > 1. **Imports all sources from the previous/original map into `self`**
        >    - `source_indexes.push(self.add_source(s));`
        > 
        > 2. **Imports all names from the previous/original map into `self`**
        >    - `names_indexes.push(self.add_name(n));`
        > 
        > 3. **Copies over `sourcesContent`**
        >    - via `self.set_source_content(...)`
        > 
        > 4. **Walks every mapping already in `self`**
        >    - `for line_content in self.inner.mapping_lines.iter_mut()`
        >    - `for mapping in line_content.mappings.iter_mut()`
        > 
        > 5. **For each mapping that already has an `original` location, it looks that location up in the previous map using `find_closest_mapping`**
        >    - `original_sourcemap.find_closest_mapping(original_location.original_line, original_location.original_column)`
        > 
        > 6. **If that found mapping itself has an `original`, it replaces `self`’s original location with that deeper original**
        >    - `*original_location_option = Some(OriginalLocation::new(...))`
        > 
        > 7. **If no mapping is found, or the found mapping has no original, it drops the original mapping**
        >    - `*original_location_option = None;`
        > 
        > That means this is not an offset merge like `addSourceMap`; it is an in-place rewrite of `self`’s `mapping.original` fields by chasing them through `original_sourcemap.find_closest_mapping(...)`.
  - https://github.com/parcel-bundler/source-map/blob/b290dae1bea9bf249a0691e06014e09f974f70dd/parcel_sourcemap/src/lib.rs#L122-L138
    - This is the rust implementation of `find_closest_mapping`

---

Based on how those implementations work (eg. `find_closest_mapping`), and looking back at `oxc_sourcemap`'s docs, I see the following seem potentially relevant:

- `lookup_token`: https://docs.rs/oxc_sourcemap/latest/oxc_sourcemap/struct.SourceMap.html#method.lookup_token
  - > Lookup a token by line and column, it will used at remapping.
- https://docs.rs/oxc_sourcemap/latest/oxc_sourcemap/struct.SourceMap.html#method.lookup_source_view_token
  - > Lookup a token by line and column, it will used at remapping. See `SourceViewToken`.

**Comment by erictheswift** [comment]
Another use case, if it helps — we hit this in a production webpack app.

Our loader chain is `oxc-webpack-loader` ← our own instrumentation loader ← `@wyw-in-js/webpack-loader` (webpack runs `use[]` right to left, so oxc goes last). Both of the earlier loaders emit a sourcemap, and both get thrown away: `oxc-webpack-loader` takes them as `_inputSourceMap` and can't pass them anywhere, since `oxc-transform` only has `sourcemap?: boolean`.

So the final map points back to whatever oxc was handed, not to the original `.ts`. Sentry frames in those modules land on the wrong lines, and the CSS-in-JS extraction hop just isn't there at all.

I only noticed because I'd added a map to our loader and wanted to confirm it was doing something. Built twice, once forwarding the map and once not — the emitted `.map` files came out byte-identical. As far as I can tell there's currently no way for a loader sitting before oxc in a webpack chain to contribute mappings.

`oxc-webpack-loader` can't really fix it on its own, it needs what you're describing here. If a prototype ever lands I'm happy to try it on a real ~7900-module build and report back.
